curp/src/sync_manager.rs

```diff
@@ -415,4 +444,225 @@ impl<C: Command + 'static> SyncManager<C> {
+            Ok(resp) => {
+                let resp = resp.into_inner();
+
+                // calibrate term
+                let state = state.upgradable_read();
+                if resp.term > state.term {
+                    let mut state = RwLockUpgradableReadGuard::upgrade(state);
+                    state.update_to_term(resp.term);
+                    return;
+                }
+
+                // is still a candidate
+                if !matches!(state.role(), ServerRole::Candidate) {
+                    return;
+                }
+
+                #[allow(clippy::integer_arithmetic)]
+                if resp.vote_granted {
+                    debug!("vote is granted by server {}", connect.addr);
+                    let mut state = RwLockUpgradableReadGuard::upgrade(state);
+                    state.votes_received += 1;
+
+                    // the majority has granted the vote
+                    let min_granted = (state.others.len() + 1) / 2 + 1;
+                    if state.votes_received > min_granted {
+                        state.set_role(ServerRole::Leader);
+                        debug!("server becomes leader");
+
+                        // init next_index
+                        let last_log_index = state.last_log_index();
+                        for index in state.next_index.values_mut() {
+                            *index = last_log_index + 1; // iter from the end to front is more likely to match the follower
+                        }
+
+                        // trigger calibrate
+                        calibrate_trigger.notify(1);
+                    }
+                }
+            }
+        }
+    }
+
+    /// Leader should first enforce followers to be consistent with it when it comes to power
+    #[allow(clippy::integer_arithmetic, clippy::indexing_slicing)] // log.len() >= 1 because we have a fake log[0], indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+    async fn leader_calibrates_followers(
+        connects: Vec<Arc<Connect>>,
+        state: Arc<RwLock<State<C>>>,
+        calibrate_trigger: Arc<Event>,
+    ) {
+        loop {
+            calibrate_trigger.listen().await;
+            for connect in connects.iter().cloned() {
+                let state = Arc::clone(&state);
+                let _handle = tokio::spawn(async move {
+                    loop {
+                        // send append entry
+                        #[allow(clippy::shadow_unrelated)] // clippy false positive
+                        let (
+                            term,
+                            prev_log_index,
+                            prev_log_term,
+                            entries,
+                            leader_commit,
+                            last_sent_index,
+                        ) = state.map_read(|state| {
+                            let next_index = state.next_index[&connect.addr];
```

Comment:
For a new leader, it usually knows nothing about the next_index of others.

Reply:
A new leader will reset all server's next_index to the index of last log first.

Reply:
It makes sense.